### ISSUE TITLE
[BUG] Cannot bind to ICloudBlob with Microsoft.Azure.WebJobs.Extensions.Storage 5.0.0

### ISSUE THREAD

**Issue by codebrane** [Azure/azure-sdk-for-net#25429]
**Describe the bug**
After upgrading Microsoft.Azure.WebJobs.Extensions.Storage from 4.0.5 to 5.0.0 and adding Microsoft.Azure.Storage.Blob 11.2.3 the function cannot bind to ICloudBlob.
It works after downgrading to 4.0.5 and removing Microsoft.Azure.Storage.Blob which is not required for 4.0.5 but is required for 5.0.0.

**Expected behavior**
The function should bind to ICloudBlob

**Actual behavior (include Exception or Stack Trace)**
The function fails to bind to ICloudBlob

```
[2021-11-19T08:23:28.449Z] Error indexing method 'BrunelWMSServiceBus'
[2021-11-19T08:23:28.450Z] Microsoft.Azure.WebJobs.Host: Error indexing method 'BrunelWMSServiceBus'. Microsoft.Azure.WebJobs.Host: Can't bind Blob to type 'Microsoft.Azure.Storage.Blob.ICloudBlob'.
[2021-11-19T08:23:28.450Z] Possible causes:
[2021-11-19T08:23:28.450Z] 1) Tried binding to 'Azure.Storage.Blobs.BlobContainerClient, Azure.Storage.Blobs, Version=12.10.0.0, Culture=neutral, PublicKeyToken=92742159e12e44c8' but user type assembly was 'Microsoft.Azure.Storage.Blob.ICloudBlob, Microsoft.Azure.Storage.Blob, Version=11.2.3.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35.
[2021-11-19T08:23:28.450Z] 2) Tried binding to 'Microsoft.Azure.WebJobs.Extensions.Storage.Blobs.Config.BlobsExtensionConfigProvider+MultiBlobContext, Microsoft.Azure.WebJobs.Extensions.Storage.Blobs, Version=5.0.0.0, Culture=neutral, PublicKeyToken=92742159e12e44c8' but user type assembly was 'Microsoft.Azure.Storage.Blob.ICloudBlob, Microsoft.Azure.Storage.Blob, Version=11.2.3.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35.
[2021-11-19T08:23:28.450Z] 3) Tried binding to 'System.IO.Stream, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e' but user type assembly was 'Microsoft.Azure.Storage.Blob.ICloudBlob, Microsoft.Azure.Storage.Blob, Version=11.2.3.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35.
[2021-11-19T08:23:28.450Z] 4) Tried binding to 'Azure.Storage.Blobs.Specialized.BlockBlobClient, Azure.Storage.Blobs, Version=12.10.0.0, Culture=neutral, PublicKeyToken=92742159e12e44c8' but user type assembly was 'Microsoft.Azure.Storage.Blob.ICloudBlob, Microsoft.Azure.Storage.Blob, Version=11.2.3.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35.
[2021-11-19T08:23:28.450Z] 5) Tried binding to 'Azure.Storage.Blobs.Specialized.PageBlobClient, Azure.Storage.Blobs, Version=12.10.0.0, Culture=neutral, PublicKeyToken=92742159e12e44c8' but user type assembly was 'Microsoft.Azure.Storage.Blob.ICloudBlob, Microsoft.Azure.Storage.Blob, Version=11.2.3.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35.
[2021-11-19T08:23:28.450Z] 6) Tried binding to 'Azure.Storage.Blobs.Specialized.AppendBlobClient, Azure.Storage.Blobs, Version=12.10.0.0, Culture=neutral, PublicKeyToken=92742159e12e44c8' but user type assembly was 'Microsoft.Azure.Storage.Blob.ICloudBlob, Microsoft.Azure.Storage.Blob, Version=11.2.3.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35.
[2021-11-19T08:23:28.450Z] 7) Tried binding to 'Azure.Storage.Blobs.BlobClient, Azure.Storage.Blobs, Version=12.10.0.0, Culture=neutral, PublicKeyToken=92742159e12e44c8' but user type assembly was 'Microsoft.Azure.Storage.Blob.ICloudBlob, Microsoft.Azure.Storage.Blob, Version=11.2.3.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35.
[2021-11-19T08:23:28.451Z] 8) Tried binding to 'Azure.Storage.Blobs.Specialized.BlobBaseClient, Azure.Storage.Blobs, Version=12.10.0.0, Culture=neutral, PublicKeyToken=92742159e12e44c8' but user type assembly was 'Microsoft.Azure.Storage.Blob.ICloudBlob, Microsoft.Azure.Storage.Blob, Version=11.2.3.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35.
```

**To Reproduce**
Upgrade Microsoft.Azure.WebJobs.Extensions.Storage from 4.0.5 to 5.0.0 and run function locally with:
`[Blob("thecontainer/thefile.json", FileAccess.ReadWrite, Connection = "StorageConnectionStringAzure")] ICloudBlob theBlob`

**Environment:**
Microsoft.Azure.WebJobs.Extensions.Storage 5.0.0
Microsoft.Azure.Storage.Blob 11.2.3
.NET runtime 3.1
VisualStudio for Mac 8.10.14 (build 7)

**Comment by codebrane** [comment]
I looked in:

`sdk/hdinsight/Microsoft.Azure.HDInsight.Job/src/Customizations/Models/AzureStorageAccess.cs`

and saw that it uses:

`ICloudBlob blobReference;`

so I called the parameter in the function `blobReference` and it worked. I then changed it back to `theBlob` and now it is working too.

Why would changing the parameter name have any effect on the binding and why would it not bind until the parameter name was changed?

It's the same situation trying to bind `MessageReceiver messageReceiver`.

`[2021-11-19T10:02:57.897Z] Microsoft.Azure.WebJobs.Host: Error indexing method 'BrunelWMSServiceBus'. Microsoft.Azure.WebJobs.Host: Can't bind parameter 'messageReceiver' to type 'Microsoft.Azure.ServiceBus.Core.MessageReceiver'.
`

I just haven't found out what the parameter needs to be called presumably.

All of these worked before upgrading to 5.0.0.
